flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkContainer.java

```diff
@@ -0,0 +1,251 @@
+        return this.restClusterClient;
+    }
+
+    /**
+     * Restarts JobManager container.
+     *
+     * <p>Note that the REST port will be changed because the new JM container will be mapped to
+     * another random port. Pleas make sure to get the REST cluster client again after this method
+     * is invoked.
+     */
+    public void restartJobManager(RunnableWithException afterFailAction) throws Exception {
+        if (this.haService == null) {
+            LOG.warn(
+                    "Restarting JobManager without HA service. This might drop all your running jobs");
+        }
+        jobManager.stop();
+        afterFailAction.run();
+        jobManager.start();
+        // Recreate client because JobManager REST port might have been changed in new container
+        this.restClusterClient = createClusterClient();
+        waitUntilAllTaskManagerConnected();
+    }
+
+    /** Restarts all TaskManager containers. */
+    public void restartTaskManager(RunnableWithException afterFailAction) throws Exception {
+        taskManagers.forEach(GenericContainer::stop);
+        afterFailAction.run();
+        taskManagers.forEach(GenericContainer::start);
+    }
+
+    /**
+     * Submits an SQL job to the running cluster.
+     *
+     * <p><b>NOTE:</b> You should not use {@code '\t'}.
+     */
+    public void submitSQLJob(SQLJobSubmission job) throws IOException, InterruptedException {
+        checkState(isRunning(), "SQL job submission is only applicable for a running cluster");
+        // Create SQL script and copy it to JobManager
+        final List<String> commands = new ArrayList<>();
+        Path script = Files.createTempDirectory("sql-script").resolve("script");
+        Files.write(script, job.getSqlLines());
+        jobManager.copyFileToContainer(MountableFile.forHostPath(script), "/tmp/script.sql");
+
+        // Construct SQL client command
+        commands.add("cat /tmp/script.sql | ");
+        commands.add("flink/bin/sql-client.sh");
+        for (String jar : job.getJars()) {
+            commands.add("--jar");
+            Path path = Paths.get(jar);
+            String containerPath = "/tmp/" + path.getFileName();
+            jobManager.copyFileToContainer(MountableFile.forHostPath(path), containerPath);
+            commands.add(containerPath);
+        }
+
+        // Execute command in JobManager
+        Container.ExecResult execResult =
+                jobManager.execInContainer("bash", "-c", String.join(" ", commands));
+        LOG.info(execResult.getStdout());
+        LOG.error(execResult.getStderr());
+        if (execResult.getExitCode() != 0) {
+            throw new AssertionError("Failed when submitting the SQL job.");
+        }
+    }
+
+    // ------------------------ JUnit 5 lifecycle management ------------------------
+    @Override
+    public void beforeAll(ExtensionContext context) throws Exception {
+        this.start();
```

Comment:
What is the purpose of defining the start() and stop() as two public methods? Will they be called in other lifecycle outside of this package or even in other modules beyond JUnit 5?

Reply:
Yes there are cases that not just use it as a JUnit 5 extension, such as in `FlinkContainerTestEnvironment`, so I kept these two public methods in the class for manually controlling the lifecycle of it.